File: retort-annotations/src/main/java/co/kaioru/retort/annotation/builder/IAnnotationFactoryBuilder.java
package co.kaioru.retort.annotation.builder;

import co.kaioru.retort.ICommand;
import co.kaioru.retort.ICommandContext;
import co.kaioru.retort.annotation.IAnnotationAdapter;

import java.util.Collection;

public interface IAnnotationFactoryBuilder<I extends ICommandContext, O> {

    IAnnotationFactoryBuilder<I, O> withAdapter(IAnnotationAdapter<I, O> adapter);

    IAnnotationFactoryBuilder<I, O> withObject(Object object);

    Collection<ICommand<I, O>> build();

}

File: retort-annotations/src/main/kotlin/co/kaioru/retort/annotation/builder/IAnnotationFactoryBuilder.kt
package co.kaioru.retort.annotation.builder

import co.kaioru.retort.ICommand
import co.kaioru.retort.annotation.IAnnotationAdapter
import co.kaioru.retort.annotation.IAnnotationFactory

interface IAnnotationFactoryBuilder<I, O> : IAnnotationFactory<I, O> {
    fun withAdapter(adapter: IAnnotationAdapter<I, O>): IAnnotationFactoryBuilder<I, O> {
        registerAdapter(adapter)
        return this
    }

    fun withAdapters(adapter: Collection<IAnnotationAdapter<I, O>>): IAnnotationFactoryBuilder<I, O> {
        registerAdapters(adapters)
        return this
    }

    fun withObject(any: Any)

    fun withObjects(any: Collection<Any>)

    fun build(): Collection<ICommand<I, O>>
}